kong/plugins/prometheus/wasmx.lua

```diff
@@ -0,0 +1,210 @@
+    local label_pos, _, _, value = str_find(pw_key, pair.pattern)
+
+    if label_pos and value then
+      table_insert(labels, { pair.label, value })
+
+      first_label_pos = (label_pos < first_label_pos) and label_pos or first_label_pos
+    end
+  end
+
+  if first_label_pos ~= #pw_key then
+    -- discarding labels from metric_name
+    metric_name = str_sub(pw_key, 1, first_label_pos - 1)
+  end
+
+  return metric_name, labels
+end
+
+
+local function parse_key(key)
+  -- TODO: parse wa. (WasmX metrics) and lua. (metrics defined in Lua land)
+  local header = { pw = "pw." }
+
+  local name = key
+  local labels = {}
+
+  local is_pw = #key > #header.pw and key:sub(0, #header.pw) == header.pw
+
+  if is_pw then
+    name, labels = parse_pw_key(key)
+  end
+
+  name = name:gsub("%.", "_")
+
+  return name, labels
+end
+
+
+local function serialize_labels(labels)
+  local buf = buf_new()
+
+  for _, pair in ipairs(labels) do
+    buf:put(fmt(',%s="%s"', pair[1], pair[2]))
+  end
+
+  buf:get(1)  -- discard trailing comma
+
+  return "{" .. buf:get() .. "}"
+end
+
+
+local function serialize_metric(m, buf)
+  buf:put(fmt("# HELP %s\n# TYPE %s %s", m.name, m.name, m.type))
+
+  if m.type == "histogram" then
+    local sum_lines_buf = buf_new()
+    local count_lines_buf = buf_new()
+
+    for _, pair in ipairs(m.labels) do
+      local count, sum = 0, 0
+      local labels, labeled_m = pair[1], pair[2]
+      local slabels = (#labels > 0) and serialize_labels(labels) or ""
+
+      local blabels = (#labels > 0) and (slabels:sub(1, #slabels - 1) .. ",") or "{"
+
+      for _, bin in ipairs(labeled_m.value) do
+        local ub = (bin.ub ~= 4294967295) and bin.ub or "+Inf"
+        local ubl = fmt('le="%s"', ub)
+
+        count = count + bin.count
+
+        buf:put(fmt("\n%s%s %s", m.name, blabels .. ubl .. "}", count))
+      end
+
+      sum = sum + labeled_m.sum
+
+      sum_lines_buf:put(fmt("\n%s_sum%s %s", m.name, slabels, sum))
+      count_lines_buf:put(fmt("\n%s_count%s %s", m.name, slabels, count))
+    end
+
+    buf:put(sum_lines_buf:get())
+    buf:put(count_lines_buf:get())
+
+  else
+    for _, pair in ipairs(m.labels) do
+      local labels, labeled_m = pair[1], pair[2]
+      local slabels = (#labels > 0) and serialize_labels(labels) or ""
+
+      buf:put(fmt("\n%s%s %s", m.name, slabels, labeled_m.value))
+    end
+  end
+
+  buf:put("\n")
+end
+
+
+_M.metric_data = function()
+  local metrics = {}
+  local parsed = {}
+  local buf = buf_new()
+
+  -- delayed require of the WasmX module, to ensure it is loaded
+  -- after ngx_wasm_module.so is loaded.
+  if not wasmx_shm then
+    local ok, _wasmx_shm = pcall(require, "resty.wasmx.shm")
+    if ok then
+      wasmx_shm = _wasmx_shm
+    end
+  end
+
+  if not wasmx_shm then
+    return
+  end
+
+  wasmx_shm.metrics:lock()
+
+  for key in wasmx_shm.metrics:iterate_keys() do
+    local pair = { key, wasmx_shm.metrics:get_by_name(key, { prefix = false })}
```

Comment:
nit: we can avoid extra table creation by caching this option table somewhere

```suggestion
    local pair = { key, wasmx_shm.metrics:get_by_name(key, GET_METRIC_OPTS)}
    -- somewhere else: `local GET_METRIC_OPTS = { prefix = false }`
```